MudAutoComplete: Fix CoerceText not working on empty search results (#2617)

BederBourahmah opened the pull request:
Fixes #2617 

A call to `ToggleMenu` is not being triggered when clicking away from the `MudAutocomplete` component. `ToggleMenu` is where the call to `CoerceTextToValue` happens. I found that `IsOpen` was being set to false whenever the search function returned an empty list, but this didn't actually do anything, from what I could tell. So, removing this one line fixes this issue.

I looked into adding it to `OnInputBlurred`, but that doesn't work because when a selection is made it also triggers `OnInputBlurred`, as the comment by @henon pointed out.)

henon commented:
Please add a testcase to our autocomplete test suite showing that it works and protecting this fix against accidental unfixing by future PRs. You can look at the other tests to see how it is done.

BederBourahmah commented:
It looks like there was already a test that was supposed to exercise CoerceText in the case of no results found, but it was not catching this bug. I figured that I just needed to add a check to make sure `IsOpen` was still true after an empty search result was returned. However, when I tested that change on master it was still passing. I found that because `OnSearch` was an `async void` method, the state was not getting updated fast enough before the check on `IsOpen` was executed in the test. I changed `OnSearch` so it returned a `Task` and added `await` in the methods that called it. After that change, the test failed on master and passed on this branch.

henon commented:
ok, your solution is fine but there is also another way. you could have done a 

```c#
comp.WaitForAssertion(()=>autocomplete.IsOpen.Should().BeTrue());
```

if I am not mistaken.

BederBourahmah commented:
I don't think `WaitForAssertion` would work here because the core problem in the unit test was that `IsOpen` was returning **true** (the passing state) when I tested on **master**, when it really should have been returning **false** because `OnSearch` was erroneously setting it to false when no results were found. The test was incorrectly seeing a pass condition for `IsOpen` value because there was still code executing within the component when it was performing the check. I actually tried `WaitForAssertion` on master and I was getting inconsistent results. Sometimes the test passed, sometimes it failed because `WaitForAssertion` timed out, and sometimes it failed because `Text` didn't match the expected on the last line. I attribute that inconsistency to a race condition set into motion by the `async void` fire-and-forget pattern that was in place. The `Task` & `await` pattern guarantees that the code is done executing by the time the test checks `IsOpen`.

henon commented:
Awesome. Thanks for the thorough explanation!
